erpnext/hr/report/monthly_attendance_sheet/monthly_attendance_sheet.py

```diff
@@ -120,137 +138,258 @@ def get_chart_data(emp_att_map, days):
+			get_leave_type_and_total_leave_taken(row, conditions, filters, emp, leave_list=leave_list)
+			get_late_entry_and_earl_exit_count(row, conditions, filters)
+
 		emp_att_map[emp] = emp_status_map
 		record.append(row)
-
 	return record, emp_att_map
 
+def get_late_entry_and_earl_exit_count(row, conditions, filters):
+	time_default_counts = frappe.db.sql("""select (select count(*) from `tabAttendance` where \
+		late_entry = 1 %s) as late_entry_count, (select count(*) from tabAttendance where \
+		early_exit = 1 %s) as early_exit_count""" % (conditions, conditions), filters)
+
+	row.extend([time_default_counts[0][0],time_default_counts[0][1]])
+
+def get_leave_type_and_total_leave_taken(row, conditions, filters, emp, leave_list=None):
+	leave_details = frappe.db.sql("""select leave_type, status, count(*) as count from `tabAttendance`\
+		where leave_type is not NULL %s group by leave_type, status""" % conditions, filters, as_dict=1)
+
+	leaves = {}
+	for d in leave_details:
+		if d.status == "Half Day":
+			d.count = d.count * 0.5
+		if d.leave_type in leaves:
+			leaves[d.leave_type] += d.count
+		else:
+			leaves[d.leave_type] = d.count
+
+
+	for d in leave_list:
+		d = d.replace(":Float:120", "")
+		print(d)
+		if d in leaves:
+			row.append(leaves[d])
+		else:
+			row.append("0.0")
+
+def get_totals(status, status_map, attendance_detail):
+
+	present = absent = leave = holiday = unmarked= 0.0
+	if status and attendance_detail:
+		if status in ["Weekly Off", "Holiday"]:
+			holiday += 1
+		if status_map[status]['is_present']:
+			present += 1
+		elif status_map[status]['is_leave']:
+			leave += 1
+		elif status_map[status]['is_half_day']:
+			leave += 0.5
+			remaining_half_day_status = attendance_detail[2]
+			if status_map[remaining_half_day_status]["is_present"]:
+				present += 0.5
+			else:
+				absent += 0.5
+		else:
+			absent +=  1
+	else:
+		unmarked += 1
+
+	return [present, leave, absent, holiday, unmarked]
+
+def get_abbr(status, status_map, leave_abbr_map, attendance_detail):
+	abbr = ''
+	abbr = status_map[status]['abbr']
+	if attendance_detail:
+		leave_type = attendance_detail[1]
+
+		if status_map[status]['is_leave'] and leave_type:
+			abbr = leave_abbr_map[leave_type]["full_day_abbr"]
+
+		if status_map[status]['is_half_day'] and leave_type:
+			abbr = leave_abbr_map[leave_type]["half_day_abbr"]
+			remaining_half_day_status = attendance_detail[2]
+
+			rem_half_day_abbr = status_map[remaining_half_day_status]['abbr']
+			abbr += " + " + rem_half_day_abbr
+	return abbr
+
+def get_status(attendance_detail, holiday_map, emp_holiday_list, day):
+	status = None
+	if attendance_detail:
+		status = attendance_detail[0]
+
+	if status is None and holiday_map:
+		if emp_holiday_list in holiday_map:
+			for idx, ele in enumerate(holiday_map[emp_holiday_list]):
+				if day == holiday_map[emp_holiday_list][idx][0]:
+					if holiday_map[emp_holiday_list][idx][1]:
+						status = "Weekly Off"
+					else:
+						status = "Holiday"
```

Comment:
Should this loop still continue running once you get that day in the holiday map? If not, you can return the status from here itself:

```suggestion
					if holiday_map[emp_holiday_list][idx][1]:
						return "Weekly Off"
					else:
						return "Holiday"
```

If you choose to do this, on line 271 also, return the status directly.